### ISSUE TITLE
Have field reference for array_column

### ISSUE THREAD

**Issue by SpyroTEQ** [kalessil/phpinspectionsea#1518]
This issue was a nice feature: https://github.com/kalessil/phpinspectionsea/issues/931

But it misses something: typehinting

Because now, if you have:

```php
class X { public $x; }
/** @var X[] */
$xs = array(new X(), new X(), new X());

var_dump(array_column($xs, 'x'));
```
 
Then you have no auto-completion, refactoring hints, or even inspection on the `'x'` magic string. Hence by doing a refactoring like this, changing `$x` to `$y`:
```php
class X { public $y; }
/** @var X[] */
$xs = array(new X(), new X(), new X());

var_dump(array_column($xs, 'x'));
```

Then you will have broken code during execution that could have been spotted with static analysis.

So, could this be handled in some way? Like either showing an error on `'x'` if `$xs` is of type `X[]` and field `x` cannot be found in `X`, or better, inject that `'x'` should be a reference to the `x` field of `X` (and so, intellij would show if field does not exist/allow refactoring/allow ctrl+click)?

**Comment by kalessil** [comment]
Adde new settings, the properties mapping is disabled by default.

**Comment by SpyroTEQ** [comment]
Oh, I wasn't thinking about adding settings about "when to run the referenced inspection". I was thinking about adding a new inspection to spot when `array_column($arrayOfObjects, 'fieldOfObjects')` references an unexisting field. Like, here, inspection would show an error on `'fieldOfObjects'` if the type of `$arrayOfObjects` if `\any\object\classname` and if `$fieldOfObjects` is not a field in this class.

I manages to implement this in my own plugin doing that:

```java
@NotNull
@Override
public PhpElementVisitor buildVisitor(
		@NotNull final ProblemsHolder problemsHolder,
		final boolean isOnTheFly) {
	return new PhpElementVisitor() {
		@Override
		public void visitPhpFunctionCall(final FunctionReference reference) {
			super.visitPhpFunctionCall(reference);

			if (!"array_column".equals(reference.getName())) {
				return;
			}

			final PsiElement[] params = reference.getParameters();

			if (params.length != 2
					|| !(params[0] instanceof PhpTypedElement)
					|| !(params[1] instanceof StringLiteralExpression)) {
				return;
			}

			final PhpTypedElement firstParam = (PhpTypedElement) params[0];
			final StringLiteralExpression secondParam = (StringLiteralExpression) params[1];
			final String fieldName = secondParam.getContents();

			final Boxed<Boolean> hasOneMatchingClass = new Boxed<>(false);
			final Boxed<Boolean> hasArrayType = new Boxed<>(false);
			firstParam.getType()
					.getTypes()
					.forEach(cname -> {
						final String fqn = cname.replace("[]", "");
						if ("\\array".equals(fqn)) {
							hasArrayType.setValue(true);
							return;
						}

						PhpIndex.getInstance(reference.getProject())
								.getClassesByFQN(fqn)
								.forEach(c -> {
									hasOneMatchingClass.setValue(true);
									if (c.getFields().stream()
											.noneMatch(f -> f.getName().equals(fieldName))) {
										ReinomInspection.problem(
												problemsHolder,
												secondParam,
												"No ->" + fieldName + " field in class " + fqn);
									}
								});
					});

			if (hasArrayType.getValue() && !hasOneMatchingClass.getValue()) {
				ReinomInspection.warning(
						problemsHolder,
						firstParam,
						"This array is not typed");
			}
		}
	};
}
```

Which results in:

```php
<?php

namespace variispace;

class ArrayColTest {
	public $yes;
	private $pvField;
}

// Puts "The array is not typed" on "$items"
$items = array();
array_column($items, 'whatever');

// Puts "The array is not typed" on "$items2"
(static function (array $items2) {
	array_column($items2, 'whatever');
})(array());

// Puts "No ->whatever field in class \variispace\ArrayColTest" as error on 'whatever'
/** @var ArrayColTest[] $items3 */
$items3 = array();
array_column($items3, 'whatever');

// Shows nothing: it's all good
array_column($items3, 'yes');

// Shows nothing (I haven't handled private fields: I guess it should be considered as erroneous?)
array_column($items3, 'pvField');
```

**Comment by kalessil** [comment]
Re-opening in order to revisit solution from @SpyroTEQ
